Assigments1

NikitaAware opened the pull request:
Thank you for contributing to NullAway!

Please note that once you click "Create Pull Request" you will be asked to sign our [Uber Contributor License Agreement](https://cla-assistant.io/uber/NullAway) via [CLA assistant](https://cla-assistant.io/).

Before pressing the "Create Pull Request" button, please provide the following:

  - [x] A description about what and why you are contributing, even if it's trivial.

  - [x] The issue number(s) or PR number(s) in the description if you are contributing in response to those.

  - [x] If applicable, unit tests.

msridhar committed:
an initial test case

msridhar committed:
sample code

NikitaAware committed:
add-generics-checks

NikitaAware committed:
first-case-of-generics-support

NikitaAware committed:
first-case-of-generics-support

msridhar committed:
restore bad deletion

NikitaAware committed:
add-generics-spport-task-1-updates

NikitaAware committed:
add-generics-spport-task-1-updates

NikitaAware committed:
add-generics-spport-task-1-updates

NikitaAware committed:
add generics check for variables

msridhar committed:
Merge branch 'master' into add-generics-checks

msridhar committed:
tweak error message

msridhar committed:
more tests

NikitaAware committed:
code cleanup.remove print statements

NikitaAware committed:
Add temporary check to consider unannoted base type

NikitaAware committed:
minor code changes

NikitaAware committed:
add generics check for new class - check if the class is generic

NikitaAware committed:
add generics check for new class - check for the type arguments

NikitaAware committed:
generics check for new class

NikitaAware committed:
generics check for extended class

NikitaAware committed:
generics check for new class

NikitaAware committed:
generics check for new class - minor changes for error message

NikitaAware committed:
generics check for interface implementation

NikitaAware committed:
minor identation changes

NikitaAware committed:
nullability checks for build failures

NikitaAware committed:
Add Test: generics check for subclasses

NikitaAware committed:
restructure generics tests

NikitaAware committed:
restructure generics tests

NikitaAware committed:
Add tests

NikitaAware committed:
restructure code

NikitaAware committed:
restructure code

NikitaAware committed:
restructure code

msridhar committed:
Merge branch 'master' into add-generics-checks

msridhar committed:
add jspecify mode flag

msridhar committed:
check jspecify flag

msridhar committed:
add test case for nested type parameter instantiation

NikitaAware committed:
nested type

NikitaAware committed:
nested generics type check  for new class

NikitaAware committed:
Add test - generics check for function return type

NikitaAware committed:
Add - generics check for function return type

NikitaAware committed:
code cleaning - generics check

NikitaAware committed:
code cleaning - generics check

msridhar committed:
Merge branch 'master' into add-generics-checks

NikitaAware committed:
Suggested changes

NikitaAware committed:
Suggested changes

NikitaAware committed:
Suggested changes

NikitaAware committed:
working changes

NikitaAware committed:
working changes

NikitaAware committed:
suggested changes

NikitaAware committed:
suggested changes

NikitaAware committed:
suggested changes

NikitaAware committed:
suggested changes

msridhar committed:
extend test

NikitaAware committed:
@Nullable annotation check for Paramterized typed tree

NikitaAware committed:
@Nullable annotation check for Paramterized typed tree

NikitaAware committed:
added updates

NikitaAware committed:
added updates

NikitaAware committed:
extra test cases for other annotations

NikitaAware committed:
extra test cases for other annotations

NikitaAware committed:
Merge branch 'master' into add-generics-checks

msridhar committed:
some comments and a todo

msridhar committed:
reuse code

msridhar committed:
minor cleanup

msridhar committed:
fix

NikitaAware committed:
Suggested changes

NikitaAware committed:
Suggested changes

NikitaAware committed:
Suggested changes

msridhar committed:
Javadoc tweaks

msridhar committed:
add private constructor

NikitaAware committed:
suggested changes

NikitaAware committed:
suggested changes

NikitaAware committed:
suggested changes

NikitaAware committed:
suggested changes

msridhar committed:
code cleanup

msridhar committed:
tests for error positions

msridhar committed:
Clean up by implementing ParameterizedTypeTreeMatcher

msridhar committed:
remove extra blank line

msridhar committed:
add test

NikitaAware committed:
suggested changes

NikitaAware committed:
suggested changes

msridhar committed:
Merge branch 'master' into add-generics-checks

msridhar committed:
tweak error message

msridhar committed:
clarify test name and purpose

NikitaAware committed:
Assignment interface added

NikitaAware committed:
structure for assignments checks

NikitaAware committed:
Assignment changes

NikitaAware committed:
supertype changes

NikitaAware committed:
Merge branch 'master' into assigments1

NikitaAware committed:
multiple interface implementation

NikitaAware committed:
multilevel inheritance

NikitaAware committed:
changes

NikitaAware committed:
changes

NikitaAware committed:
nested checks

NikitaAware committed:
working changes

NikitaAware committed:
working changes

NikitaAware committed:
Working changes

NikitaAware committed:
updates

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -90,4 +119,162 @@ private static void invalidInstantiationError(
+
+  @SuppressWarnings("UnusedVariable")
+  public static void checkInstantiationForAssignments(
+      AssignmentTree tree, Config config, VisitorState state, NullAway analysis) {
```
We will need to generalize this check to more than `AssignmentTree`s; e.g., we need to handle initializers in `VariableTree`s.  To do so we may need to generalize this method's parameters, or create another method for `VariableTree`s.  For a `VariableTree` `t` (e.g., `Foo<@Nullable String> f = new Foo<@Nullable String>();`), we should do a check if `t.getInitializer()` is not `null`.  Then, the LHS type comes from `t.getType()`.  For the cases we care about this will likely be a `ParameterizedTypeTree`; we'll have to get if the generic type annotations are properly present.

NikitaAware committed:
failing 3 tests

NikitaAware committed:
working tests

NikitaAware committed:
npe for variable initialization changes

NikitaAware committed:
npe for variable initialization changes

NikitaAware committed:
All tests pass including the existing tests

msridhar committed:
Merge branch 'master' into assigments1

msridhar commented on nullaway/src/main/java/com/uber/nullaway/NullAway.java:
```diff
@@ -1326,6 +1330,16 @@ public Description matchVariable(VariableTree tree, VisitorState state) {
+    // assigned to null
+    if (tree.getInitializer() != null) {
+      if (state.getSourceForNode(tree.getInitializer()) != null) {
+        if (!state.getSourceForNode(tree.getInitializer()).equals("null")) {
```
Why do we need this check?  In any case we shouldn't be checking for `null` by getting the source code, there is surely a better way

msridhar commented on nullaway/src/test/java/com/uber/nullaway/NullAwayJSpecifyGenericsTests.java:
```diff
@@ -198,6 +198,160 @@ public void downcastInstantiation() {
+            " static void testOKOtherAnnotation(NullableTypeParam<String> t) {",
+            "       NullableTypeParam<String> t3;",
+            "       // BUG: Diagnostic contains: Generic type parameter",
+            "        t3 = new NullableTypeParam<@Nullable String>();",
```
Can we just use our `VariableTree` support now rather than having this assignment be separate?  Same for other similar examples

msridhar replied:
This still needs to be updated to use `VariableTree` support

msridhar commented on nullaway/src/test/java/com/uber/nullaway/NullAwayJSpecifyGenericsTests.java:
```diff
@@ -198,6 +198,160 @@ public void downcastInstantiation() {
+            "       NullableTypeParamMultipleArguments<String, String> t5 = new NullableTypeParamMultipleArguments<String, String>();",
+            "       NullableTypeParamMultipleArguments<@Nullable String, String> t6 = new NullableTypeParamMultipleArguments<@Nullable String, String>();",
+            "       // BUG: Diagnostic contains: Generic type parameter",
+            "       t5 = t6;",
```
This is fine to keep as an assignment

msridhar commented on nullaway/src/test/java/com/uber/nullaway/NullAwayJSpecifyGenericsTests.java:
```diff
@@ -198,6 +198,160 @@ public void downcastInstantiation() {
+            "  void sampleError() {",
+            "    SuperClassC<@Nullable String> f;",
+            "    // BUG: Diagnostic contains: Generic type parameter",
+            "    f = new FnImpl1();",
```
Again use `VariableTree` support

msridhar commented on nullaway/src/test/java/com/uber/nullaway/NullAwayJSpecifyGenericsTests.java:
```diff
@@ -198,6 +198,160 @@ public void downcastInstantiation() {
+            "  void sampleValidInstantiation() {",
+            "  SuperClassC<@Nullable String> f;",
+            "    // No error",
+            "    f = new FnImpl2();",
```
Again use `VariableTree` support

msridhar commented on nullaway/src/test/java/com/uber/nullaway/NullAwayJSpecifyGenericsTests.java:
```diff
@@ -198,6 +198,160 @@ public void downcastInstantiation() {
+            "package com.uber;",
+            "import org.jspecify.annotations.Nullable;",
+            "class Test {",
+            "  class A<P1 extends @Nullable Object, P2 extends @Nullable Object>{}",
```
Looks like the `A` class is unused?

msridhar commented on nullaway/src/test/java/com/uber/nullaway/NullAwayJSpecifyGenericsTests.java:
```diff
@@ -198,6 +198,160 @@ public void downcastInstantiation() {
+  }
+
+  @Test
+  public void NestedVariableDeclarationChecks() {
```
Can probably delete this test and just use `VariableTree` support elsewhere

msridhar committed:
Merge branch 'master' into assigments1

msridhar committed:
some cleanup

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -4,21 +4,54 @@
+  public static Type supertypeMatchingLHS(
+      Type.ClassType lhsType, Type.ClassType rhsType, VisitorState state) {
+    // interfaces
+    List<Type> listOfDirectSuperTypes = rhsType.all_interfaces_field;
```
We can clean up this code a bunch by calling `state.getTypes.closure(rhsType)`, which will give you a list of all supertypes, including both superclasses and interfaces.

msridhar committed:
enable check for multiple top-level classes

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -90,4 +123,177 @@ private static void invalidInstantiationError(
+  }
+}
+
+class ParameterizedTypeTreeNullableArgIndices
```
We should not have multiple top-level classes in the same source file.  These should be `static` nested classes of `GenericsChecks`.  I enabled the Error Prone check that catches this.

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -90,4 +123,177 @@ private static void invalidInstantiationError(
+    return nullableTypeArgIndices;
+  }
+
+  public static void superTypeMatchingRHSParameterizedTypeTree(
```
Rather than making this a static method, have the enclosing object take the relevant `rhs` into the constructor and store the state in a field.  Then, this method can just take maybe the other parameters.  Similar comment for `checkAssignmentTypeMatch`, and same for the methods in `NormalTypeTreeNullableTypeArgIndices`

msridhar reviewed (commented):
This is starting to look pretty good!  I can understand the high-level logic now.  Here are some cleanup suggestions.

NikitaAware committed:
suggested changes - enclosing class member fields

NikitaAware committed:
code cleanup

NikitaAware committed:
suggested changes

msridhar committed:
Merge branch 'master' into assigments1

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -4,19 +4,45 @@
+      Type.ClassType lhsType, Type.ClassType rhsType, VisitorState state) {
+    // all supertypes including classes as well as interfaces
+    List<Type> listOfDirectSuperTypes = state.getTypes().closure(rhsType);
+    if (listOfDirectSuperTypes != null) {
```
Can this really be null?  Shouldn't we only call this method in cases where `rhsType` is a subtype of `lhsType` that is not equal to `lhsType`?

NikitaAware replied:
instead of checking if it is a direct supertype first and then calling this function I am just calling the function and checking it here. I am not sure what is the logic behind checking whether it is the direct super type or not. considering it iterates through the list of the super types and checks whether it is a direct supertype and then again checks in the same list for the direct supertype matching the lhs type, it will check the same thing twice.

msridhar commented on nullaway/src/main/java/com/uber/nullaway/AnnotatedTypeWrapper.java:
```diff
@@ -0,0 +1,11 @@
+import com.sun.source.tree.Tree;
+import java.util.HashSet;
+
+public interface AnnotatedTypeWrapper<T1, T2> {
```
So I think the overall code structure should look like this roughly:
```java
void checkIdenticalInstantiations(Tree lhsTree, Tree rhsTree, ...) {
  TypeWrapper lhsWrapper = getTypeWrapper(lhsTree);
  TypeWrapper rhsWrapper = getTypeWrapper(rhsTree);
  checkIdenticalWrappers(lhsWrapper, rhsWrapper);
}

void checkIdenticalWrappers(TypeWrapper lhsWrapper, TypeWrapper rhsWrapper) {
  if (lhsWrapper == null || rhsWrapper == null) return;
  if (!ASTHelpers.isSameType(lhsWrapper.getWrapped(), rhsWrapper.getWrapped()) {
    rhsWrapper = rhsWrapper.viewAsSupertype(lhsWrapper.getWrapped());
  }
  if (!sameNullableTypeIndices(lhsWrapper,rhsWrapper)) reportError(...);
  List<TypeWrapper> lhsNested = lhsWrapper.getWrappersForNested();
  List<TypeWrapper> rhsNested = rhsWrapper.getWrappersForNested();
  for (int i = 0; i < lhsNested.size(); i++) {
    checkIdenticalWrappers(lhsNested.get(i), rhsNested.get(i));
  }
}

@Nullable TypeWrapper getTypeWrapper(Tree tree) {
  // logic based on the type of the tree, and return null if
  // it's not for a generic type or doesn't need to be checked
}

interface TypeWrapper {
  // the underlying type being wrapped
  Type getWrapped();
  List<Integer> getNullableTypeIndices();
  List<TypeWrapper> getWrappersForNested(); // can have null elements
}
```

The key differences from your current code structure are:

* The `AnnotatedTypeWrapper` functions should need to take in the thing to be wrapped as an argument.  It should just be passed to the constructor and stored in a field.
* The logic to check for an assignment type match should be written **once** outside the type wrappers, not twice, once inside each type wrapper.  This way we avoid duplication of logic.  That checking logic should rely on the `AnnotatedTypeWrapper` APIs for all logic related to the types, which is why in the example above I added a method `getWrappersForNested()` to get the nested type wrappers.

Does this make sense?  If so please refactor the code to match this structure.

NikitaAware committed:
some suggested changes

NikitaAware committed:
suggested changes with working normal typed tree

NikitaAware committed:
Changes with parameterized typed tree without the error generation

NikitaAware committed:
suggested changes with two test cases failing

NikitaAware committed:
suggested changes with all passing tests

NikitaAware committed:
All working changes

NikitaAware committed:
Continuous integration working

NikitaAware committed:
Continuous integration working

NikitaAware committed:
changes

NikitaAware committed:
minor changes

msridhar committed:
fix naming case

msridhar committed:
failing test

msridhar committed:
refactoring and comments

msridhar committed:
add a case

NikitaAware committed:
working code with class type

NikitaAware committed:
minor refactoring

NikitaAware committed:
minor changes

NikitaAware committed:
code cleaning

NikitaAware committed:
code cleaning

msridhar committed:
fancy test works!

msridhar committed:
Merge branch 'master' into assigments1

msridhar committed:
only run check in JSpecify mode

msridhar committed:
Fix NullAway errors

NikitaAware committed:
modified assignment message

msridhar committed:
remove unneeded suppression

msridhar committed:
remove change to build.gradle

msridhar committed:
remove local used only once

msridhar committed:
cleanup supertypeMatchingLHS

msridhar committed:
restore private

msridhar committed:
add private

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -1,22 +1,64 @@
 
-  private GenericsChecks() {
-    // just utility methods
+  private static final String NULLABLE_NAME = "org.jspecify.annotations.Nullable";
```
Eventually we may need to handle `@Nullable` type-use annotations other than jspecify.  but we can leave this for now

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -90,4 +132,141 @@ private static void invalidInstantiationError(
+    ErrorBuilder errorBuilder = analysis.getErrorBuilder();
+    ErrorMessage errorMessage =
+        new ErrorMessage(
+            ErrorMessage.MessageTypes.TYPE_PARAMETER_CANNOT_BE_NULLABLE,
```
We should create a new message type for this, like `ErrorMessage.MessageTypes.ASSIGN_GENERIC_NULLABLE`

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -90,4 +132,141 @@ private static void invalidInstantiationError(
+    ErrorMessage errorMessage =
+        new ErrorMessage(
+            ErrorMessage.MessageTypes.TYPE_PARAMETER_CANNOT_BE_NULLABLE,
+            String.format(
```
This message isn't quite right.  Maybe we can pass in the full (original) lhs and rhs types, and then write "Cannot assign from type " + rhsType + " to type " + lhsType + " due to mismatched nullability of type parameters"

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -90,4 +132,141 @@ private static void invalidInstantiationError(
+          typeWithPreservedAnnotations(
+              (ParameterizedTypeTree) ((NewClassTree) rhsTree).getIdentifier());
+
+      ParameterizedTypeTree paramTypedTree =
+          (ParameterizedTypeTree) ((NewClassTree) rhsTree).getIdentifier();
+      // not generic
+      if (paramTypedTree.getTypeArguments().size() <= 0) {
+        return;
+      }
```
This check and return should happen first, before re-assigning `rhsType`

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -90,4 +132,141 @@ private static void invalidInstantiationError(
+  }
+
+  private void compareAnnotations(Type lhsType, Type rhsType, Tree tree) {
+    if (lhsType instanceof Type.ClassType && rhsType instanceof Type.ClassType) {
+      rhsType = supertypeMatchingLHS((Type.ClassType) lhsType, (Type.ClassType) rhsType);
+    }
```
Let's change the type of `lhsType` and `rhsType` parameters to be `ClassType` rather than doing this check; the method makes no sense if the types passed in are not `ClassType`s

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -90,4 +132,141 @@ private static void invalidInstantiationError(
+    }
+    List<Type> lhsTypeArguments = lhsType.getTypeArguments();
+    List<Type> rhsTypeArguments = rhsType.getTypeArguments();
+    // this error should already be generated by some other part of NullAway still adding a check to
+    // be safe
+    if (lhsTypeArguments.size() != rhsTypeArguments.size()) {
+      return;
+    }
```
We should throw a `RuntimeException` if this check fails rather than just returning; it's a violation of a fundamental assumption we are making

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -90,4 +132,141 @@ private static void invalidInstantiationError(
+        invalidAssignmentInstantiationError(tree, lhsAnnotation, rhsAnnotation, state, analysis);
+        return;
+      }
+      // nested generics
+      if (lhsTypeArguments.get(i).getTypeArguments().length() > 0) {
+        compareAnnotations(lhsTypeArguments.get(i), rhsTypeArguments.get(i), tree);
+      }
```
👍

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -90,4 +132,141 @@ private static void invalidInstantiationError(
+        JCTree.JCAnnotatedType annotatedType = (JCTree.JCAnnotatedType) typeArguments.get(i);
+        for (JCTree.JCAnnotation annotation : annotatedType.getAnnotations()) {
+          Attribute.Compound attribute = annotation.attribute;
+          if (attribute.toString().equals("@org.jspecify.annotations.Nullable")) {
```
Let's store `"@org.jspecify.annotations.Nullable"` in a static final field, computed from the existing `NULLABLE_TYPE` field

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -90,4 +132,141 @@ private static void invalidInstantiationError(
+      // nested generics checks
+      Type currentArgType = ASTHelpers.getType(typeArguments.get(i));
+      if (currentArgType != null && currentArgType.getTypeArguments().size() > 0) {
+        Type.ClassType nestedTyp =
+            typeWithPreservedAnnotations((ParameterizedTypeTree) typeArguments.get(i));
+        newTypeArgs.add(nestedTyp);
```
Hrm, it looks like the `myAnnos` list is not used at all in this case.  Don't we want to do the recursive call, and then take the result and still do `cloneWithMetadata` like in the `else` case?  We should add a test that fails with the current (buggy) logic

msridhar commented on nullaway/src/main/java/com/uber/nullaway/NullAway.java:
```diff
@@ -1326,6 +1330,12 @@ public Description matchVariable(VariableTree tree, VisitorState state) {
       return Description.NO_MATCH;
     }
     VarSymbol symbol = ASTHelpers.getSymbol(tree);
+    // working on this. This part is needed to avoid assignment checks where the variable is
+    // assigned to null
```
Delete the comment

msridhar reviewed (commented):
Looking really good!  Several cleanup comments and one possible bug

NikitaAware committed:
changed the error message

NikitaAware committed:
updated the error message text

NikitaAware committed:
Type.ClassType suggestion changes

NikitaAware committed:
changes

NikitaAware committed:
changes

NikitaAware committed:
changes

NikitaAware committed:
extra test case

msridhar committed:
add failing test

NikitaAware committed:
updated logic

NikitaAware committed:
minor updates

NikitaAware committed:
minor code cleaning

NikitaAware committed:
minor code cleaning

NikitaAware committed:
suggested changes

NikitaAware committed:
suggested changes

NikitaAware committed:
replaced super type matching lhs with asSuper method

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -1,22 +1,45 @@
-    // just utility methods
+  private static final String NULLABLE_NAME = "org.jspecify.annotations.Nullable";
+
+  private static final String NULLABLE_TYPE = "@org.jspecify.annotations.Nullable";
```
```suggestion
  private static final String NULLABLE_ANNOTATION = "@" + NULLABLE_NAME;
```

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -90,4 +113,148 @@ private static void invalidInstantiationError(
+    List<Type> lhsTypeArguments = lhsType.getTypeArguments();
+    List<Type> rhsTypeArguments = rhsType.getTypeArguments();
+    if (lhsTypeArguments.size() != rhsTypeArguments.size()) {
+      throw new RuntimeException(rhsType + " does not match " + lhsType);
```
```suggestion
      throw new RuntimeException("number of types arguments in " + rhsType + " does not match " + lhsType);
```

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -90,4 +113,148 @@ private static void invalidInstantiationError(
+    }
+    for (int i = 0; i < lhsTypeArguments.size(); i++) {
+      com.sun.tools.javac.util.List<Attribute.TypeCompound> annotationMirrorsLHS =
+          lhsTypeArguments.get(i).getAnnotationMirrors();
```
Store `lhsTypeArguments.get(i)` in a local variable `curLhsTypeArg` and re-use throughout loop

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -90,4 +113,148 @@ private static void invalidInstantiationError(
+      com.sun.tools.javac.util.List<Attribute.TypeCompound> annotationMirrorsLHS =
+          lhsTypeArguments.get(i).getAnnotationMirrors();
+      com.sun.tools.javac.util.List<Attribute.TypeCompound> annotationMirrorsRHS =
+          rhsTypeArguments.get(i).getAnnotationMirrors();
```
Store `rhsTypeArguments.get(i)` in a local variable `curRhsTypeArg` and re-use throughout loop

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -90,4 +113,148 @@ private static void invalidInstantiationError(
+        Type.ClassType nestedTyp =
+            typeWithPreservedAnnotations((ParameterizedTypeTree) typeArguments.get(i));
+        Type.ClassType nestedTypArg = castToNonNull(nestedTyp).cloneWithMetadata(metaData);
+        newTypeArgs.add(nestedTypArg);
```
remove code duplication with lines 251-252

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -90,4 +113,148 @@ private static void invalidInstantiationError(
+    for (int i = 0; i < typeArguments.size(); i++) {
+      List<Attribute.TypeCompound> allAnnotations = new ArrayList<>();
+      List<JCTree.JCAnnotation> annotations = new ArrayList<>();
+      if (typeArguments.get(i).getClass().equals(JCTree.JCAnnotatedType.class)) {
```
Use `AnnotatedTypeTree` instead of the internal `JCAnnotatedType`, and use `instanceof` instead of `getClass().equals(...)`

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -90,4 +113,148 @@ private static void invalidInstantiationError(
+      if (typeArguments.get(i).getClass().equals(JCTree.JCAnnotatedType.class)) {
+        JCTree.JCAnnotatedType annotatedType = (JCTree.JCAnnotatedType) typeArguments.get(i);
+        annotations = annotatedType.getAnnotations();
+      } else if (typeArguments.get(i) instanceof JCTree.JCTypeApply
```
Use `ParameterizedTypeTree` instead of `JCTypeApply`

NikitaAware replied:
Working on this.

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -90,4 +113,148 @@ private static void invalidInstantiationError(
+      List<JCTree.JCAnnotation> annotations = new ArrayList<>();
+      if (typeArguments.get(i).getClass().equals(JCTree.JCAnnotatedType.class)) {
+        JCTree.JCAnnotatedType annotatedType = (JCTree.JCAnnotatedType) typeArguments.get(i);
+        annotations = annotatedType.getAnnotations();
```
Just assign the `annotatedType` variable in the `if` condition, and then declare this variable and assign it after the `if`, removing the duplication with line 229

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -90,4 +113,148 @@ private static void invalidInstantiationError(
+    List<Type> newTypeArgs = new ArrayList<>();
+    for (int i = 0; i < typeArguments.size(); i++) {
+      List<Attribute.TypeCompound> allAnnotations = new ArrayList<>();
+      List<JCTree.JCAnnotation> annotations = new ArrayList<>();
```
Use `AnnotationTree` instead of internal `JCAnnotation`

NikitaAware replied:
Working on this comment. Somehow typer arguments are not instances of the Annotation Tree here. Will check and update this.

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -90,4 +113,148 @@ private static void invalidInstantiationError(
+      for (JCTree.JCAnnotation annotation : annotations) {
+        Attribute.Compound attribute = annotation.attribute;
+        if (attribute.toString().equals(NULLABLE_TYPE)) {
+          allAnnotations.add(
+              new Attribute.TypeCompound(nullableType, com.sun.tools.javac.util.List.nil(), null));
```
`allAnnotations` should only ever have 0 or 1 elements, right?  If I understand correctly, then get rid of `allAnnotations`.  Instead, just track a boolean `hasNullableAnnotation`.  Then below, conditionally assign `annos`: if `hasNullableAnnotation` is true, assign `com.sun...List.from(Collections.singletonList(new Attribute.TypeCompound(...)))`, otherwise assign `com.sun...List.nil()`

msridhar commented on nullaway/src/test/java/com/uber/nullaway/NullAwayJSpecifyGenericsTests.java:
```diff
@@ -198,6 +198,200 @@ public void downcastInstantiation() {
 
+  // assignment tree tests
+  @Test
+  public void genericsChecksForAssignments() {
```
Generally, we need to clean up these tests.  It should be relatively obvious to a reader what the purpose of each line is.  So we should split up the tests into more methods as needed, give them meaningful names, and remove redundant tests

msridhar reviewed (commented):
Looking really good and close now!  Here are a few more comments to address

NikitaAware committed:
updated assignment checks

NikitaAware committed:
changes

NikitaAware committed:
changes

NikitaAware committed:
suggested changes

NikitaAware committed:
suggested changes

NikitaAware committed:
suggested changes

NikitaAware committed:
suggested changes

NikitaAware committed:
suggested changes

msridhar committed:
Merge branch 'master' into assigments1

msridhar committed:
fix nullaway error

msridhar committed:
Merge branch 'master' into assigments1

msridhar committed:
fixes

msridhar committed:
more cleanup and javadoc

msridhar committed:
Merge branch 'master' into assigments1

msridhar committed:
Merge branch 'master' into assigments1

XN137 committed:
Support more test assertions in OptionalEmptinessHandler (#718)

supported with JUnit 4:
```
Assert.assertTrue(x.isPresent())
Assert.assertFalse(x.isEmpty())
```

supported with JUnit 5:
```
Assertions.assertTrue(x.isPresent())
Assertions.assertFalse(x.isEmpty())
```

supported with AssertJ:
```
assertThat(x.isPresent()).isTrue()
assertThat(x.isEmpty()).isFalse()

assertThat(x).isPresent()
assertThat(x).isNotEmpty()
```

supported with Truth:
```
assertThat(x.isPresent()).isTrue()
assertThat(x.isEmpty()).isFalse()
```

msridhar committed:
Update to Error Prone 2.18.0 (#707)

Fixed some minor Javadoc warnings reported by the new version.  Also, adds an `@Inject` annotation on the main `NullAway` constructor, for compatibility with future versions of Error Prone that load checkers via Guice.

msridhar committed:
Merge branch 'master' into assigments1

NikitaAware committed:
suggested changes in test nestedChecksForAssignmentsMultipleArguments

NikitaAware committed:
suggested changes in test - superTypeAssignmentChecksSingleInterface

NikitaAware committed:
suggested indentation changes in test - superTypeAssignmentChecksMultipleInterface

NikitaAware committed:
suggested changes

NikitaAware committed:
suggested changes

NikitaAware committed:
suggested test for the lambda expressions

NikitaAware committed:
Java doc for type with preserved annotations

NikitaAware committed:
comments

NikitaAware committed:
code cleaning for the tests

NikitaAware committed:
code cleaning for the tests

NikitaAware committed:
changes

NikitaAware committed:
suggested changes

msridhar commented:
Closing as this is now an upstream PR
